### PR TITLE
[7601] Install and configure Celery

### PR DESCRIPTION
**Pull request by hklarner**

### CODE REVIEW

File: requirements/base.txt
Comment (m4ra) [inline requirements/base.txt:32]: should be moved to a4 since a4 has also background task
see: https://github.com/liqd/adhocracy-plus/blob/main/requirements/base.txt#L16-L20

**Review by m4ra — CHANGES_REQUESTED** [review]

File: apps/__init__.py
Comment (m4ra) [inline apps/__init__.py:5]: So we can move this part to `adhocrac-plusy/__init__.py` and do:
```
from .config.celery import app as celery_app

__all__ = ('celery_app',)
```
then make use of `from celery import shared_task` decorator instead of importing the configured app. then we don't have issues with the hyphen.
ref: https://docs.celeryq.dev/en/stable/django/first-steps-with-django.html#using-celery-with-django

File: adhocracy-plus/config/settings/base.py
Comment (hklarner) [reply adhocracy-plus/config/settings/base.py:116]: done

File: adhocracy-plus/config/celery.py
Comment (hklarner) [reply adhocracy-plus/config/celery.py:14]: done

File: requirements/base.txt
Comment (hklarner) [reply requirements/base.txt:32]: done, see https://github.com/liqd/adhocracy4/pull/1488

File: apps/__init__.py
Comment (hklarner) [reply apps/__init__.py:5]: done

File: docs/celery.md
Comment (hklarner) [reply docs/celery.md:16]: done

File: docs/celery.md
Comment (hklarner) [reply docs/celery.md:11]: done

File: docs/celery.md
Comment (hklarner) [reply docs/celery.md:21]: done

File: Makefile
Comment (m4ra) [inline Makefile:210]: good to add `--loglevel INFO` after the worker

File: Makefile
Comment (m4ra) [inline Makefile:48]: it is running celery in stdout (foreground) not in the background.
Also to be consistent with the above postgres make command, I'd put `start/stop/status` as a prefix rather than a suffix.

**Review by m4ra — CHANGES_REQUESTED** [review]
Looks very good!
just few tiny corrections

File: adhocracy-plus/config/settings/base.py
Comment (hklarner) [reply adhocracy-plus/config/settings/base.py:562]: good catch, changed it to `CELERY_RESULT_BACKEND = "redis://localhost:6379"

File: adhocracy-plus/config/settings/base.py
Comment (hklarner) [reply adhocracy-plus/config/settings/base.py:114]: good catch. done.

File: Makefile
Comment (hklarner) [reply Makefile:210]: agreed, done.

File: Makefile
Comment (goapunk) [reply Makefile:48]: @hklarner I think @m4ra was referring to the explanation of the command, it still says `-- starts the celery worker in the background` which is wrong. Changing `start-postgres` to `postgres-start` is fine with me, both ways work

**Review by goapunk — CHANGES_REQUESTED** [review]

File: docs/celery.md
Comment (m4ra) [reply docs/celery.md:30]: Oops, the README still is missing how to start/stop/status Celery with your make command.

File: adhocracy-plus/config/celery.py
Comment (hklarner) [reply adhocracy-plus/config/celery.py:13]: done. I have also extended the `celery-worker-status` command to call the dummy task, fetch the result from redis and print it to the screen. See [`pull request review changes 4`](https://github.com/liqd/adhocracy-plus/pull/2523/commits/74d824d1007d41f3cdb97e27953095cf0a35b1d8).

File: Makefile
Comment (hklarner) [reply Makefile]: done

File: README.md
Comment (m4ra) [inline README.md:48]: can you add the steps for working with celery? still missing in the README

File: docs/celery.md
Comment (m4ra) [inline docs/celery.md:46]: these here can be moved/copied in the README, plus the celery status command, along with the other make commands.

File: apps/projects/tasks.py
Comment (m4ra) [inline apps/projects/tasks.py:7]: can we rename object to object_dict or use typing hint? @goapunk I think we should do that while we touch legacy code, is fine if only some parts has it, it will grow slowly. Already talked about it breifly with @hklarner

File: README.md
Comment (hklarner) [reply README.md:48]: done

File: docs/celery.md
Comment (hklarner) [reply docs/celery.md:46]: the stop command has actually disappeared, since start is now in the foreground. But it was still in the makefile help text. I remoevd it.

done.

File: tests/conftest.py
Comment (m4ra) [inline tests/conftest.py:41]: I wouldn't create another celery here for a+, but I'd add the celery_config in the fixtures:
https://docs.celeryq.dev/en/stable/userguide/testing.html#session-scope

Comment (m4ra) [reply tests/conftest.py:41]: Agreed to leave this option open for future

**Review by m4ra — COMMENTED** [review]
background task need the migration to zero as we discussed, then we can remove them from settings + requirements

File: Makefile
Comment (goapunk) [inline Makefile:215]: should this go into a separate command? I find it a bit weird/unexpected behavior that asking for the status also executes a task

Comment (hklarner) [reply Makefile:215]: done

**Review by m4ra — APPROVED** [review]

File: adhocracy-plus/config/settings/base.py
Comment (m4ra) [inline adhocracy-plus/config/settings/base.py:39]: oops this needs to go back for the time being.

**Review by m4ra — CHANGES_REQUESTED** [review]

File: adhocracy-plus/config/settings/base.py
Comment (hklarner) [reply adhocracy-plus/config/settings/base.py]: good catch, done.

File: docs/installation_prod.md
Comment (goapunk) [inline docs/installation_prod.md:91]: lets recommend unix sockets for a prod install

File: requirements/base.txt
Comment (goapunk) [inline requirements/base.txt:15]: I'm unsure if this should only be in `prod.txt`, during development we don't need it as we want always_eager anyway. What do you think @hklarner @m4ra

File: adhocracy-plus/config/settings/base.py
Comment (goapunk) [inline adhocracy-plus/config/settings/base.py:560]: If we add them here testing locally will produce an error every time a celery task is used. We have to add `CELERY_TASK_ALWAYS_EAGER = True` to `dev.py` so it doesn't require a broker/backend.

**Review by goapunk — CHANGES_REQUESTED** [review]
looks good, just some small things

File: requirements/base.txt
Comment (hklarner) [reply requirements/base.txt:15]: makes sense. i'll change it.

File: adhocracy-plus/config/settings/base.py
Comment (hklarner) [reply adhocracy-plus/config/settings/base.py:560]: done

**Review by goapunk — CHANGES_REQUESTED** [review]

File: adhocracy-plus/config/settings/base.py
Comment (m4ra) [reply adhocracy-plus/config/settings/base.py:562]: Yes, False shall be the default. In tests, @hannes is configuring it to True.

Comment (hklarner) [reply adhocracy-plus/config/settings/base.py:562]: great catch! you're right, the `CELERY_TASK_ALWAYS_EAGER = True` is not overwritten in `local,py` in production!
done.

**Review by goapunk — APPROVED** [review]
nice, looks good to me! needs a4 update and then good to go

File: requirements/base.txt
Comment (goapunk) [inline requirements/base.txt:3]: unnecessary newline here :)

**Review by goapunk — CHANGES_REQUESTED** [review]

**Review by goapunk — APPROVED** [review]

File: adhocracy-plus/config/settings/dev.py
Comment (goapunk) [inline adhocracy-plus/config/settings/dev.py:52]: probably good to add a comment here saying it should always be at the bottom

Comment (hklarner) [reply adhocracy-plus/config/settings/dev.py:40]: done.

**Comment by hklarner** [comment]
I have added a new celery configuration parameter:
```
CELERY_RESULT_EXTENDED = True
```
that instructs celery to save more information in the results backend. So what used to be:
```
$ redis-cli GET celery-task-meta-8b3be356-9e50-4c41-8b62-ec766504bf04 | jq
{
  "status": "SUCCESS",
  "result": null,
  "traceback": null,
  "children": [],
  "date_done": "2023-09-19T12:33:11.118701",
  "task_id": "8b3be356-9e50-4c41-8b62-ec766504bf04"
}
```
is now
```
$ redis-cli GET celery-task-meta-467e1aff-a7c6-4fb8-aa47-093aa86150d8 | jq
{
  "status": "SUCCESS",
  "result": null,
  "traceback": null,
  "children": [],
  "date_done": "2023-09-20T07:53:10.285548",
  "name": "adhocracy4.emails.tasks.send_async",
  "args": [],
  "kwargs": {
    "email_module_name": "apps.notifications.emails",
    "email_class_name": "NotifyCreatorOnModeratorFeedback",
    "app_label": "a4_candy_ideas",
    "model_name": "idea",
    "object_pk": 62,
    "args": [],
    "kwargs": {}
  },
  "worker": "celery@tuxedo86",
  "retries": 0,
  "queue": "celery",
  "task_id": "467e1aff-a7c6-4fb8-aa47-093aa86150d8"
}
```

**Comment by hklarner** [comment]
As discussed with goapunk, I have added `redis == 5.0.0` back to `requirements/dev.txt` to avoid this attribute error when developing locally:
```
AttributeError: 'NoneType' object has no attribute 'Redis'
```

File: apps/projects/tasks.py
Comment (m4ra) [inline apps/projects/tasks.py:7]: Lets pls rename it to something else.. send_email_data or anything other than this.

File: apps/notifications/emails.py
Comment (m4ra) [inline apps/notifications/emails.py:193]: related to a4 celery [PR](https://github.com/liqd/adhocracy4/pull/1493), `object_dict` is not necessary here, so I suggest to update the kwargs instead of defining a new dict `object_dict`.
And I would definitely rename the task to get rid of `...no_object` in it.

File: apps/projects/tasks.py
Comment (goapunk) [reply apps/projects/tasks.py:7]: I think the idea is to replace this one with the refactored one in a4, right?

File: apps/projects/tasks.py
Comment (hklarner) [reply apps/projects/tasks.py]: as discussed today we will postpone any real refactoring to another PR and focus on making celery work. I have there simply removed the return statement.

File: apps/projects/tasks.py
Comment (hklarner) [reply apps/projects/tasks.py]: as discussed today we will postpone any real refactoring to another PR and focus on making celery work. I have there simply removed the return statement.

File: apps/notifications/emails.py
Comment (hklarner) [reply apps/notifications/emails.py:193]: as discussed today we will postpone any real refactoring to another PR and focus on making celery work. I have there simply removed the return statement.

File: requirements/prod.txt
Comment (goapunk) [inline requirements/prod.txt:4]: sorry, as we discussed this should either go back to `base.txt` or the `CELERY_` settings need to move from `base.py` to `prod.py`

**Review by goapunk — CHANGES_REQUESTED** [review]

File: requirements/prod.txt
Comment (hklarner) [reply requirements/prod.txt]: done

File: docs/celery.md
Comment (m4ra) [inline docs/celery.md:22]: should be `CELERY_TASK_ALWAYS_EAGER`

File: README.md
Comment (m4ra) [inline README.md:57]: Add here what you have in docs/celery.md about celery eager should be false in local settings. Otherwise the worker won't do anything

**Review by m4ra — CHANGES_REQUESTED** [review]

File: docs/celery.md
Comment (hklarner) [reply docs/celery.md:22]: good catch, done.

File: README.md
Comment (hklarner) [reply README.md:57]: good point, have added missing instructions.

**Review by m4ra — APPROVED** [review]
Looks ready to merge once we merge a4 PR and update the requirements here

**Comment by hklarner** [comment]
> Looks ready to merge once we merge a4 PR and update the requirements here

I have replaced the last version bump commit with one that imports the latest a4, namely [db03f3](https://github.com/liqd/adhocracy4/commit/db03f3478931a39013fa50de183d5042a6cc284b).

**Review by goapunk — APPROVED** [review]
LGTM, nice work!

**Review by m4ra — APPROVED** [review]

**Pull request merged**
